Template aliasing - RFC

evs-chris opened the pull request:
__Probably post 0.8~~, not ready for merge, and no tests yet~~__

This is one of those things that pops up now and then as a little bit surprising: setting context with `with` isn't really bidirectional (#1353). There has been a proposed syntax to address that `{{#with foo.bar.baz as v1, lah.dee.dah as v2}}`, and this is my attempt to see what it would take and how it would work. What this does is introduce a new step in parsing sections, a new node type, and another step in ambiguous reference resolution to achieve aliasing within a branch of a template.

### Construction

The parsing step sits between block name and expression parsing. It looks for aliases in the form `[expression] as [non-path ref]` with optional additional `,` separated aliases if the section type is a `with`. If aliases are found, they are stuck on the template node in a new `z` property and the type is changed to `ALIAS` (19). Children are still parsed, but most of special stuff, like looking for `else`, is skipped.

The new node, `Alias`, is pretty simple. It only exists to resolve its own references and supply an `aliases` map to children. This is my first new node type, so I'm not positive that all of the methods are right. It seems to be well behaved in the stuff I've thrown through it though.

The resolution step is pretty simple. During the walk up the fragment tree, it just checks for fragment owners with an `aliases` property with a key matching the root of the current ref. If there is one, the model stored there is joined if possible. If not, resolution continues as usual.

Any valid reference is usable, so you can alias `@index` and other special refs or a regular keypath or an expression.

### Segue - components

Something that also pops up every now and then with components is how to let events fire in the context of the owner while the content guts stay in the context of the component (recently #2206). Obviously bubbling is one good solution, but that requires a little bit tighter integration between component-er and component-ee. Table that for a second while I segue the segue.

I was thinking that perhaps it would be useful to introduce a new special reference (yay!) `@ractive` that returned `this.ractive` when resolved. This could be useful in some scenarios, like the question about where to put functions that popped up recently (#2176). Then things like `@ractive.someFun(...)` could be possible. It might also be useful for combining method and proxy event syntax at some point in the future.

Now, what if template aliasing allowed you to override the `@ractive` special ref? The event directive could look for a `@ractive` override before using its own `ractive` to fire an event. This could allow something like:
```html
Name: <input value="{{current.name}}" /><br/>
Phone: <input value="{{current.phone}}" />
<List values="{{people}}">
  <div class="person" on-click="link(@keypath, 'current')">{{.name}} {{.phone}}</div>
</List>
```
and the List component:
```html
{{#each page(values)}}
  {{#with @ractive.parent as @ractive}}{{>content}}
{{/each}}
{{#each range(1, values.length / pageSize)}}
  <a href="#" on-click="set('page', .)">{{.}}</a>
{{/each}}
```
Here, the `link` call in the `content` partial would happen in the context of the component's parent, since it has an alias defined. The pagination link `set` calls would happen in the context of the component because there is no alias overriding the `@ractive` reference.

### So...

If you made it this far, thanks! I would love to get some feedback, as always, on concept and construction ( and testing when this grows some tests ).

Rich-Harris commented on src/config/types.js:
```diff
@@ -14,6 +14,7 @@ export const COMPONENT         = 15;
 export const YIELDER           = 16;
 export const INLINE_PARTIAL    = 17;
 export const DOCTYPE           = 18;
+export const ALIAS             = 19;
```
WE'RE RUNNING OUT OF NUMBERS! HALP!

evs-chris replied:
I heard CostCo has a sale on 100s on black Friday.

Rich-Harris commented:
Strongly in favour of this. The two-way binding stuff is really cool (whatever those pesky Flux-heads think), but honestly I'm probably more excited about stuff like `{{#each people as person}}` which is going to be such a big win for template readability.

Implementation-wise (and forgive the armchair quarterbacking! I haven't delved into this nearly as far as I'd like), is there a reason to have an `Alias` node instead of the `Section` managing its aliases the way it does with index/key references (i.e. passing them down to child fragments)?

I don't totally understand the `@ractive` stuff, but I think that's just because it's been a long day. :zzz:

evs-chris commented:
> I'm probably more excited about stuff like {{#each people as person}}

Totally forgot about that one! I'll have to add it...

> forgive the armchair quarterbacking

By all means! Direction is the main reason I opened this as an incomplete PR :)

>  is there a reason to have an Alias node instead of the Section managing its aliases the way it does with index/key references (i.e. passing them down to child fragments)

Well, that's the direction I started in, but it seemed easier to have a SSOT for the aliases, resolving them once in the parent and letting further resolution look up the nearest alias. Since the aliases don't really bind like indices and keys, it also didn't seem like registering and tracking them would make much sense. Though, to be fair, I haven't done any serious testing yet. I decided to go with a separate node when I started adding a bunch of conditional logic in `Section` and `Mustache`, which `Section` extends, to keep alias blocks from trying to bind to a model and act like a conditional. Once the model is out, there are bunch of places that start throwing because they assume that if they're rendered there's a model.

Also, I haven't checked yet, but on my list of things to look at is whether expressions get the same resolver when they appear more than once in a template, especially nested within another expression. So would this actually be an optimisation (proper spelling, eh)?

From
```
{{#if calculateSomethingExpensive(.) }}{{ makeStringy(calculateSomethingExpensive(.)) }}{{/if}}
````
To
```
{{#with calculateSomethingExpensive(.) as thing }}{{#if thing }}{{ makeStringy(thing) }}{{/if}}
```

> I don't totally understand the @ractive stuff

Not that it will overcome sleep deprivation, but the tl;dr is allow access to the instance in the template and allow aliasing to change which instance is returned inside alias block to allow more granular control for components.

evs-chris commented:
I've added iterative aliasing along with a few basic tests. I think I may have accidentally offended phantom at some point, because it pretty regularly refuses to succeed on tests that happily pass in a normal browser.

In an attempt to clarify my construction rambling one post up:

I introduced `Alias` to avoid extensive conditionals around model existence in `Sections` (and `Mustache`s) because an alias block has no expression to bind and it provides no context. It's just a lookup. There is now alias handling built directly onto `{{#each}}` sections, since while conceptually the same as a `{{#with}}` alias, they are entirely different beasts under the hood (pelt?).

I elected not to copy aliases around like index refs because index refs are bound to a single model that is special-case updated when stuff gets shuffled and rebound. Copying the aliases around seemed like it would add a bunch of unnecessary bookkeeping, and the only other alternative I can think of (another special model that proxies its upstream) was a bit more complexity than seemed necessary. Though, I may have confused myself about how other refs work now.

evs-chris commented:
I got a few minutes to test out computation aliases being used to share chunks of a computation, and it does behave as I expected. In edge right now, the computation is called twice, and in this branch, it is only called once.

Rich-Harris commented:
this is very cool. no reason not to include it in 0.8

![lebowski](https://cloud.githubusercontent.com/assets/1162160/10872585/729da6f4-80d3-11e5-8995-24a073699ccb.gif)

heavyk commented:
SWEEEEET!!! congrats @evs-chris !!

matthew-dean commented:
Now, combine that with declarative components FTW.

```
{{#component List { values } }}
  {{#each page(values)}}
    {{#with @ractive.parent as @ractive}}{{>content}}
  {{/each}}
  {{#each range(1, values.length / pageSize)}}
    <a href="#" on-click="set('page', .)">{{.}}</a>
  {{/each}}
{{/component}}

<List values="{{people}}">
  <div class="person" on-click="link(@keypath, 'current')">{{.name}} {{.phone}}</div>
</List>
```
Something like that would be amazing super-sauce.
